### PR TITLE
Bring the enhanced structure of the holy WebGui back into the main project

### PR DESCRIPTION
**Pull request by LunDev**
# And let everyone feel it's newer glory
## _please ignore everything before this : )_

@MyRobotLab/crack-dev-team I want you to take a look as it will replace a rather internal gear.
If you are happy with it, just hit the merge button, if not I'm fine with you changing and improving it. A reject would be a shame (in my opinion), but I will try to deal with it : )
## To sum up:
- the Webgui now lives in src/resource/Webgui/app and not in src/resource/WebGui anymore
- this is because it is now manged by npm and bower
- and (soon) to be tested by karma, jasmine & protractor (infrastructure is ready, missing the tests :P )
- they even generate reports, that will become usefull when tests are added
- and could run tests in background while you are developing to blame you instantly should you break something :P
- libaries don't have to be stored in the myrobotlab repository anymore,
- but are downloaded/updated when executing the WebGui in an IDE (eclipse, NetBeans, IntelliJ IDEA (not tested)
- or fetched and bundled by travis for the download "myrobotlab.jar"
- while reducing it's size by only including needed files
- but you still have the complete download (including the docs and examples during development!)
- this is because bower thankfully took over this responsibility and updating a libary is now as easy as editing a line (or running a command)
## Cons

Without it, it wouldn't be great - would it ?
It requires **NodeJS** - or to be more specifc npm, **N**ode's **p**ackage **m**anager
I think this isn't a big concern, as you should have node anyway - but still, it's a downside.
Just as you require Java & Git (an IDE would be usefull, too) you would need NodeJS to use the WebGui during development. (And alternative would be to manually copy dependencies over, altough I don't think this a good way. (Ignoring that this is what I've been doing with the Java dependencies for NetBeans everytime somebody updated them :P ) )
## Tested

I personally tested it on different Windows laptops (IDE and .jar) and would say that it should be worky on all Windows devices that have NodeJS installed in its default location ( C:\Program Files\nodejs ).
I'm not really sure that it worky on Linux and Macs as well (don't think so), but don't know them deep enough to make them worky. (And have no powerfull enough computer that runs Linux currently.)
## You test

I can only encourage you to try and test it out, it should be as simple as making sure NodeJS is installed (or installing it), checkout the "WebGui-Restructuring" branch and run it : )
## Conclusion

Take your time, think about it, and then maybe leave your feedback here.
Ask questions, criticize it and hate down every line of it.

As this is probably (or should I say hopefully : ) ) read by the @MyRobotLab/crack-dev-team , I want to emphasize the amazing thing you build (especially during the time I branched out), there were a lot of commits in this time. I don't know if anyone noticed, but the develop branch in the myrobotlab repository has nearly reached **2.000** commits since MyRobotLab came to GitHub! (This pull request would add another 26 : ) )
Great work everyone - and try to keep it up! _10.000 commits_, here we go! (soon)

### CODE REVIEW

**Comment by kwatters** [comment]
Very interesting,  I like the idea of having better dependency management.  I am a bit concerned about the new dependency...   will you need NodeJS installed to run the webgui, or is the dependency only to run a build?  If the later, we need to make sure TravisCI could handle it.

**Comment by LunDev** [comment]
You'll need NodeJS when you want to run the WebGui from an IDE (Eclipse, NetBeans, etc.).
Travis installs and uses it already (L11 in travis.yml)
```
node_js: "4.4.7"
```

**Comment by wilcovantoorn** [comment]
Love it!

Op 24 aug. 2016 21:00 schreef "LunDev" <notifications@github.com>:

**Comment by supertick** [comment]
This is exciting MaVo ! ..  Going to start checking it out now

**Comment by supertick** [comment]
Hee,
I think I need a little instruction on how to 'see' it in action.
For example, what are the appropriate steps now to develop it ?
We should be using bower not npm correct? to manage dependencies ?
My very basic understanding is bower is typically used for dependency
management of front-ends, very similar to how mvn & maven central work with
Java dependencies.

Packaging will use bower - but will all the packages be compressed in the
release of a new build of myrobotlab.jar ?  (in other words - the runtime
environment should not need node nor bower .. correct ?)
It also would not be necessary to have node nor bower for development of
new service gui's .. unless there are new library dependencies needed ?
 correct?

It looks like you've merged the latest develop branch service guis & js
back into the web restructuring branch ?  Is that correct ?

[image: Inline image 1]
What about current libraries ?  I think I just did a pull, and these
libraries were here, some are critical for any part of the webgui to work.
I would appreciate clarification on how one would appropriately add a new
dependency.

This is very cool stuff MaVo, I'm about to test Arduino service ui .. and
also curious of the difficulty of your process so far to merge develop
branch into your branch.
I'm hoping it was easy, as this would show we are working on a good overall
design. But how many service gui's have been tested at this point ?

....   I'll keep updating as I test

Amazing Work Marvin !

**Comment by supertick** [comment]
Appears to be an error connecting in Arduino service ui - but its in
develop too ... at least its consistent :P

[image: Inline image 1]

**Comment by supertick** [comment]
I checked in a fix for connecting on the develop branch .. but after
uploading the latest - now I get "unknown serial event 40" .. which is
completely unrelated to webgui refactoring, yet blocking me from further
testing :P
Anyone else get same results ?

[image: Inline image 1]

**Comment by LunDev** [comment]
@wilcovantoorn
>Love it!

Me too !

@supertick 
>For example, what are the appropriate steps now to develop it ?

I'm sorry, could you rephrase that?

>We should be using bower not npm correct? to manage dependencies ?

We actually use both, bower exclusivly for the libaries the webgui directly uses and npm for managing everything else, e.g. unit-testing, packaging, etc.

>My very basic understanding is bower is typically used for dependency management of front-ends, very similar to how mvn & maven central work with Java dependencies.

Jup, altough there are some guys, who use npm for everything, even for their "frontend-libaries", I like the seperation of concern (another (for me important) difference is that npm uses tree-like deps and bower resolves everything to a flat list).

>Packaging will use bower - but will all the packages be compressed in the release of a new build of myrobotlab.jar ? (in other words - the runtime environment should not need node nor bower .. correct ?)

Travis is in charge of assembling myrobotlab.jar, by letting npm install everything (including a custom npm-script that only puts needed libary-files in the jar. Therefore node and bower remain on travis and are not put inside the jar (the user doesn't need it either).

>It also would not be necessary to have node nor bower for development of new service gui's .. unless there are new library dependencies needed ? correct?

Well, currently you need just node to run the webgui as a developer (no .jar). What you could do is to copy the generated libary-dir out of a finished jar (or from somebody else) and put it in your WebGui dir. This is because node installs them when you try to execute the webgui - the big advantage is that the libaries don't have to be stored in git anymore. Personally I think having npm installed is useful for other things and tools anyway (like a quick webserver anywhere you want (http-server++ ) ), and see it as the "Java of web-developers"-the "thing" that runs your application and makes it work (don't quote that) - but that's only my opinion.

>It looks like you've merged the latest develop branch service guis & js back into the web restructuring branch ? Is that correct ?

Yes, all guis on this branch are replaced by the ones on the develop branch.

>[image: Inline image 1]

Sorry, can't view it.

>What about current libraries ? I think I just did a pull, and these libraries were here, some are critical for any part of the webgui to work. I would appreciate clarification on how one would appropriately add a new dependency.

I moved the major ones already to be installed by bower ("bower_components" and "libgen"), every libary thats not managed by bower (yet) is still in "lib". You can also still use the "old way" for adding a libary (just drop it into "lib") as bower will probably be new for many. It can be as simple as "bower install <dependency>" - but only as long as there aren't any conflicting libaries (which is quite unusual in our big project). Putting a text file in there with clear instruction would probably be a good idea.

>This is very cool stuff MaVo, I'm about to test Arduino service ui .. and also curious of the difficulty of your process so far to merge develop branch into your branch. I'm hoping it was easy, as this would show we are working on a good overall design. But how many service gui's have been tested at this point ?

develop branch ~~is~~ *was* completly merged *before your last commit*. I think merging in git is in most situations quite easy anyway - as long as you know what you are doing and can decide what to keep (e.g. I'm unhappy with a few html errors in your improved nav.html, but **love** how it looks, so I put your version in and renamed my file to keep it for reference (want to improve on your design when this pull request is done) ).
If possible I would suggest anyway that after pushing out the next release, work on a new thing/feature or big overhauls (more than one commit) are done on a seperate branch and then merged when ready (most are probably fine without out pull request). Simple fixes (1 commit with just a few changing lines) would be fine if done directly on develop.
I didn't test every single service gui, I tried to compare the behavior of service guis I don't have to buy additional components for against their counterparts on develop. They shouldn't be different anyway as their files ~~are~~ *were* synced up with develop, only things in the background changed, not the guis themselves.

> [Other two comments]

Sorry, I still can't view your pictures and don't have any input on your "unknown serial event 40", but if you want I can sync this branch up to develop again.

**Comment by MatsMRL** [comment]
Hi GroG
Make sure that you have the latest version of MRLComm uploaded. I had some
similar strange problems earlier, and uploading a new version of MRLComm
solved it.
/Mats

2016-08-26 17:14 GMT+02:00 GroG <notifications@github.com>:

> I checked in a fix for connecting on the develop branch .. but after
> uploading the latest - now I get "unknown serial event 40" .. which is
> completely unrelated to webgui refactoring, yet blocking me from further
> testing :P
> Anyone else get same results ?
>
> [image: Inline image 1]
>

**Comment by supertick** [comment]
Yes !  .. We definitely want to pull in the new glory of refactoring !
Now its just a matter of timing.

Lets do it after our long awaited release.  In a way this comes at a
perfect time as it is excellent motivation to finally get a release out.
Kevin and I discussed last night on how we can get this done, and he came
up with a great idea of making a few video tutorials on how to get started
with MRL.
When you do a video tutorial going through from downloading the jar,
loading Arduino and controlling a servo from a web page - you will *quickly
find the parts which need immediate attention*.
These parts we fix with code changes (or fix in "documentation" of the
video) - then we release !!!

After the release of MRL and tutorial videos it becomes very easy to
validate any large refactoring (Like the new Glory WebGui structure).
*The videos become a baseline for any future releases.*

I'm excited this is finally going to happen.

Now which video tutorials.  I was thinking the following.

   - Something with no hardware at all (there are potentially a large
   number of people who only have their laptops).  So a tutorial of how
   someone might do a cool activity without any hardware would be great.
   Perhaps ProgramAB, WebKit Speech Recognition & Acapela Speech.  Where you
   can talk to your (hardware-less) robot and it will talk back, perhaps even
   learn.  (Siri-ish)
   - Simple Arduino Oscope - the very basics, of how to use the MRL
   Arduino.  Upload the code, press a button, see the trace, perhaps stick a
   wire in it and see the change from the webgui.
   - Classic Arduino control a servo - similar to the one above in the
   beginning, but move a servo in the end

Kevin suggested the 'default' behavior be to start WebGui (AND) the Swing
GUIService, which I have changed.  I personally would prefer to 'only'
start the webgui by default, but I know there are still some deficiencies.
They both currently by 'default'.

I have stepped through the process of taking the latest jar and trying to
get a servo to work.  It's '*close*' - but I think there is an issue, when
you add a Servo and the Arduino is not connected ... I did this a couple
times and I think it put it in a condition of not being able to attach the
servo even after it was connected.  (Great test no ?)  It's so rare I test
things "in the jar" "end to end" :)  I'll continue to look into this and do
more testing.

I'll also start looking into how to do the merge of the "develop" branch
into "master"..

I don't see the purpose of trying to resolve the possible many conflicts on
master..
My plan is to delete master then commit develop branch onto master. (future
releases might be pull request/merge)
I bring this up again because its necessary to do this for a release, but
in the past I think you advised against it.  Let me know your current
opinion.

Now we can start doing the videos.
Kevin did a great example for the old MRL
https://www.youtube.com/watch?v=AJasOtkCbro for our older version
Alessandro did the "first" great example here -
https://www.youtube.com/watch?v=k8gfLJM3sOk

Now we need 3+ videos of our current code base !

(Oh .. regarding "For example, what are the appropriate steps now to
develop it ?")
We need a set of instructional steps posted, of how "*As a developer*" I
would do a git clone of myrobotlab & the repo - then load node and bower,
then go to webgui directory and do *bower install*
Meaning we need a good set of instructions to work with the new Glory ;)

Let's do it !

**Comment by supertick** [comment]
@Mats - it was operator error .. I set the board type as a Uno but it was a Mega :P 
The surprising thing is that it uploaded - the not surprising thing is that it didn't behave correctly.  Calamity at one point put in code in the arduino to auto-magically set the board type (it was communicated from MRLComm to the Arduino service)  .. I thought this could be potentially a problem, but now I'm starting to reconsider ;)

**Comment by LunDev** [comment]
@supertick 
I'm fine with your suggested roadplan, I'll see if I can help with videos, but probably not during the next 5-7 weeks.

> I'll also start looking into how to do the merge of the "develop" branch into "master".. 
I don't see the purpose of trying to resolve the possible many conflicts on master.. 
My plan is to delete master then commit develop branch onto master. (future releases might be pull request/merge) 
I bring this up again because its necessary to do this for a release, but in the past I think you advised against it. Let me know your current opinion.

In the past you always suggested "delete branch master" or "delete the contents of branch master". I didn't (and still don't) like(d) these two options.
My suggestion would be to replace the content of master with the content of develop using a git merge that ignores everything that is already on master and just replaces it with develop.
Speaking git commands:
```
git checkout develop
git merge -s ours master
git checkout master
git merge develop
```
and after executing these 4 commands, the content of master mirrors exactly the content of develop.

For a future developing system (after the next release), I would still suggest a system similar to this one -> http://nvie.com/posts/a-successful-git-branching-model/ that would need to be documented. During this change you may even cut down the access rights of everyone except you to certain branches (optional).

> (Oh .. regarding "For example, what are the appropriate steps now to develop it ?") We need a set of instructional steps posted, of how "*As a developer*" I would do a git clone of myrobotlab & the repo - then load node and bower, then go to webgui directory and do *bower install* Meaning we need a good set of instructions to work with the new Glory ;)

Installing NodeJs is enough, everything else (including bower and npm install) is auto-magically taken care of.

Speaking your words: "I would do a git clone of myrobotlab & the repo - then load node ~~and bower, then go to webgui directory and do *bower install*~~"

**Comment by supertick** [comment]
LunDev certainly had the right idea, however, this commit is over 2 years old and now lots has changed.
I'm revisiting this for 2 reasons 

- original webgui app should have been put in a subdirectory as typical of angular web apps
- i want to begin another webapp in react

The nodejs should not be a requirement of building, however, I'm excited by the possibility of a polyglot myrobotlab - where native services could be developed against a myrobotlab-core.js.
Potentially, if done right I believe this webapp could be shared between any mrl native service whether it be java, nodejs, python  .... etc.  The contract between the UI and the backend is a message definition over websockets, and any of the languages I mentioned can adhere to such a contract.

I have done a little research into react, and the most complicated UI elements they have as simple extensible pluggble modules.  I also like some of react's design principles vs angular.

With the correct directory structure - multiple frameworks can be quickly evaluated without the need to "replace" one

Our Angular app (1.5) is effectively "ancient" as Angular was completely re-written in version 2 and is already at version 7?  

Another reason to deal with this now, is the Jenkins multi-branch - i want all branches to be building successfully, and this one currently isn't

**Comment by supertick** [comment]
I've moved the WebGui client content to WebGui/app which I believe was one of the intents.  Additionally, the source I've zipped and is reachable - here - will be removing web refactoring branch - although it can still be reached as download here 
https://drive.google.com/file/d/19717imBu0zUvFAktc5qk9K8hdrN2Dxzf/view?usp=sharing

Although its 12K commits behind develop and the multi-branch build is trying to build it ...

I will remove this branch now
